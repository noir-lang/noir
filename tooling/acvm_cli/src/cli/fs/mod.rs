pub(super) mod inputs;
pub(super) mod witness;
