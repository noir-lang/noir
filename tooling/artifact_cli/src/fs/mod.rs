pub mod artifact;
pub mod inputs;
pub mod witness;
