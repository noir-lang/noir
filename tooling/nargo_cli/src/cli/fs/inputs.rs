use noirc_abi::{
    input_parser::{Format, InputValue},
    Abi, InputMap, MAIN_RETURN_NAME,
};
use std::{collections::BTreeMap, path::Path};

use crate::errors::FilesystemError;

/// Returns the circuit's parameters and its return value, if one exists.
/// # Examples
///
/// ```ignore
/// let (input_map, return_value): (InputMap, Option<InputValue>) =
///   read_inputs_from_file(path, "Verifier", Format::Toml, &abi)?;
/// ```
pub(crate) fn read_inputs_from_file<P: AsRef<Path>>(
    path: P,
    file_name: &str,
    format: Format,
    abi: &Abi,
) -> Result<(InputMap, Option<InputValue>), FilesystemError> {
    if abi.is_empty() {
        return Ok((BTreeMap::new(), None));
    }

    let file_path = path.as_ref().join(file_name).with_extension(format.ext());
    if !file_path.exists() {
        return Err(FilesystemError::MissingTomlFile(file_name.to_owned(), file_path));
    }

    let input_string = std::fs::read_to_string(file_path).unwrap();
    let mut input_map = format.parse(&input_string, abi)?;
    let return_value = input_map.remove(MAIN_RETURN_NAME);

    Ok((input_map, return_value))
}
