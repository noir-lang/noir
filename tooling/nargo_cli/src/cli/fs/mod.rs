use std::{
    fs::File,
    io::Write,
    path::{Path, PathBuf},
};

pub(super) mod inputs;
pub(super) mod program;
pub(super) mod witness;

pub(super) fn create_named_dir(named_dir: &Path, name: &str) -> PathBuf {
    std::fs::create_dir_all(named_dir)
        .unwrap_or_else(|_| panic!("could not create the `{name}` directory"));

    PathBuf::from(named_dir)
}

pub(super) fn write_to_file(bytes: &[u8], path: &Path) -> String {
    let display = path.display();

    let mut file = match File::create(path) {
        Err(why) => panic!("couldn't create {display}: {why}"),
        Ok(file) => file,
    };

    match file.write_all(bytes) {
        Err(why) => panic!("couldn't write to {display}: {why}"),
        Ok(_) => display.to_string(),
    }
}
