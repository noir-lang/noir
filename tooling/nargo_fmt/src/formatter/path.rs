use noirc_frontend::{
    ast::{Path, PathKind, UnresolvedType},
    token::{Keyword, Token},
};

use super::Formatter;

impl<'a> Formatter<'a> {
    pub(super) fn format_path(&mut self, path: Path) {
        self.skip_comments_and_whitespace();

        match path.kind {
            PathKind::Plain => (),
            PathKind::Crate => {
                self.write_keyword(Keyword::Crate);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Dep => {
                self.write_keyword(Keyword::Dep);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Super => {
                self.write_keyword(Keyword::Super);
                self.write_token(Token::DoubleColon);
            }
        }

        for (index, segment) in path.segments.into_iter().enumerate() {
            if index > 0 {
                self.write_token(Token::DoubleColon);
            }
            self.write_identifier(segment.ident);

            if let Some(generics) = segment.generics {
                self.format_turbofish(generics);
            }
        }
    }

    pub(super) fn format_turbofish(&mut self, generics: Vec<UnresolvedType>) {
        self.write_token(Token::DoubleColon);
        self.write_token(Token::Less);
        for (index, typ) in generics.into_iter().enumerate() {
            if index > 0 {
                self.write_comma();
                self.write_space();
            }
            self.format_type(typ);
        }

        // Skip trailing comma, if any
        self.skip_comments_and_whitespace();
        if self.is_at(Token::Comma) {
            self.bump();
        }

        self.write_token(Token::Greater);
    }
}
