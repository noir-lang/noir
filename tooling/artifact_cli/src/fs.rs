use std::path::Path;

use color_eyre::eyre::{self, Context};
use noirc_artifacts::{contract::ContractArtifact, program::ProgramArtifact};

/// A parsed JSON build artifact.
#[derive(Debug, Clone)]
pub enum Artifact {
    Program(ProgramArtifact),
    Contract(ContractArtifact),
}

impl Artifact {
    /// Try to parse an artifact as a binary program or a contract
    pub fn read_from_file(path: &Path) -> eyre::Result<Artifact> {
        let json = std::fs::read(path)
            .with_context(|| format!("failed to read artifact from '{}'", path.display()))?;

        let as_program = || serde_json::from_slice::<ProgramArtifact>(&json).map(Artifact::Program);
        let as_contract =
            || serde_json::from_slice::<ContractArtifact>(&json).map(Artifact::Contract);

        Ok(as_program().or_else(|e| as_contract().map_err(|_| e))?)
    }
}
