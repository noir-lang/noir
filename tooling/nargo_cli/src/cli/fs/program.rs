use std::path::{Path, PathBuf};

use nargo::artifacts::{contract::ContractArtifact, program::ProgramArtifact};
use noirc_frontend::graph::CrateName;

use crate::errors::FilesystemError;

use super::{create_named_dir, write_to_file};

pub(crate) fn save_program_to_file<P: AsRef<Path>>(
    program_artifact: &ProgramArtifact,
    crate_name: &CrateName,
    circuit_dir: P,
) -> PathBuf {
    let circuit_name: String = crate_name.into();
    save_build_artifact_to_file(program_artifact, &circuit_name, circuit_dir)
}

pub(crate) fn save_contract_to_file<P: AsRef<Path>>(
    compiled_contract: &ContractArtifact,
    circuit_name: &str,
    circuit_dir: P,
) -> PathBuf {
    save_build_artifact_to_file(compiled_contract, circuit_name, circuit_dir)
}

fn save_build_artifact_to_file<P: AsRef<Path>, T: ?Sized + serde::Serialize>(
    build_artifact: &T,
    artifact_name: &str,
    circuit_dir: P,
) -> PathBuf {
    create_named_dir(circuit_dir.as_ref(), "target");
    let circuit_path = circuit_dir.as_ref().join(artifact_name).with_extension("json");

    write_to_file(&serde_json::to_vec(build_artifact).unwrap(), &circuit_path);

    circuit_path
}

pub(crate) fn read_program_from_file<P: AsRef<Path>>(
    circuit_path: P,
) -> Result<ProgramArtifact, FilesystemError> {
    let file_path = circuit_path.as_ref().with_extension("json");

    let input_string =
        std::fs::read(&file_path).map_err(|_| FilesystemError::PathNotValid(file_path))?;
    let program = serde_json::from_slice(&input_string)
        .map_err(|err| FilesystemError::ProgramSerializationError(err.to_string()))?;

    Ok(program)
}
