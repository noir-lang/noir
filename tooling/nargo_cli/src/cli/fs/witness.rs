use std::path::{Path, PathBuf};

use acvm::acir::native_types::WitnessStack;
use nargo::constants::WITNESS_EXT;

use super::{create_named_dir, write_to_file};
use crate::errors::FilesystemError;

pub(crate) fn save_witness_to_dir<P: AsRef<Path>>(
    witness_stack: WitnessStack,
    witness_name: &str,
    witness_dir: P,
) -> Result<PathBuf, FilesystemError> {
    create_named_dir(witness_dir.as_ref(), "witness");
    let witness_path = witness_dir.as_ref().join(witness_name).with_extension(WITNESS_EXT);

    let buf: Vec<u8> = witness_stack.try_into()?;

    write_to_file(buf.as_slice(), &witness_path);

    Ok(witness_path)
}
