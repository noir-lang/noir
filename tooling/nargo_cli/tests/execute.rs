#[allow(unused_imports)]
#[cfg(test)]
mod tests {
    // Some of these imports are consumed by the injected tests
    use assert_cmd::prelude::*;
    use predicates::prelude::*;

    use std::collections::BTreeMap;
    use std::fs;
    use std::path::PathBuf;
    use std::process::Command;

    use super::*;

    test_binary::build_test_binary_once!(mock_backend, "../backend_interface/test-binaries");

    // include tests generated by `build.rs`
    include!(concat!(env!("OUT_DIR"), "/execute.rs"));
}
