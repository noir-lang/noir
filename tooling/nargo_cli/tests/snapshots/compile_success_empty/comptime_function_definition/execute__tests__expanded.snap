---
source: tooling/nargo_cli/tests/execute.rs
expression: expanded_code
---
use std::meta::type_of;

pub struct Foo {
    x: Field,
    field: Field,
}

pub fn foo(
    w: i32,
    y: Field,
    Foo { x, field: some_field }: Foo,
    mut a: bool,
    (b, c): (i32, i32),
) -> i32 {
    let _: (i32, Field, Field, Field, bool, i32, i32) = (w, y, x, some_field, a, b, c);
    1
}

#[test]
#[deprecated]
fn some_test() {}

comptime fn function_attr(f: FunctionDefinition) {
    let parameters: [(Quoted, Type)] = f.parameters();
    assert(parameters.len() == 5);
    assert(parameters[0].0 == quote { w });
    assert(parameters[1].0 == quote { y });
    assert(
        parameters[2].0
            == quote {
        Foo {
            x, field: some_field
        }
    },
    );
    assert(parameters[3].0 == quote { mut a });
    assert(parameters[4].0 == quote { (b, c) });
    let an_i32: i32 = 0;
    assert(parameters[0].1 == type_of(an_i32));
    assert(parameters[1].1 == type_of(0));
    assert(parameters[2].1 == type_of(Foo { x: 0, field: 1 }));
    assert(parameters[3].1 == type_of(true));
    assert(parameters[4].1 == type_of((an_i32, an_i32)));
    assert(f.return_type() == type_of(an_i32));
    assert(f.name() == quote { foo });
    assert(f.has_named_attribute("function_attr"));
}

comptime fn check_named_attribute(f: FunctionDefinition) {
    assert(f.has_named_attribute("test"));
    assert(f.has_named_attribute("deprecated"));
}

fn add_one(x: Field) -> Field {
    x + 1
}

comptime fn mutate_add_one(f: FunctionDefinition) {
    assert(f.parameters().len() == 0);
    f.set_parameters(&[(quote { x }, type_of(0))]);
    assert(f.parameters().len() == 1);
    assert(f.return_type() == type_of(()));
    f.set_return_type(type_of(0));
    assert(f.return_type() == type_of(0));
    f.set_body(quote { x + 1 }.as_expr().unwrap());
}

fn main() {
    assert(add_one(41) == 42);
}

contract some_contract {
    pub fn bar() -> pub Field {
        1
    }
}

comptime fn set_pub_return(f: FunctionDefinition) {
    f.set_return_public(true);
}

mod foo {
    pub fn some() {}

    comptime fn attr(f: FunctionDefinition) {
        assert(f.module().name() == quote { foo });
        assert(!f.is_unconstrained());
        f.set_unconstrained(true);
        assert(f.is_unconstrained());
        f.set_unconstrained(false);
        assert(!f.is_unconstrained());
    }
}

mod test_as_typed_expr_1 {
    pub fn method<T>(x: T) -> T {
        x
    }

    comptime fn foo(module: Module) -> Quoted {
        let method: FunctionDefinition = module.functions().filter(|f: FunctionDefinition| -> bool {
            f.name() == quote { method }
        })[0];
        let func: TypedExpr = method.as_typed_expr();
        quote {
            pub fn bar() -> i32 {
                func(1)
            }
        }
    }

    pub fn bar() -> i32 {
        method(1)
    }

    pub fn test() {
        ()
    }
}

mod test_as_typed_expr_2 {
    unconstrained fn method<T, let N: u32>(xs: [T; N]) -> u32 {
        xs.len()
    }

    comptime fn foo(module: Module) -> Quoted {
        let method: FunctionDefinition = module.functions().filter(|f: FunctionDefinition| -> bool {
            f.name() == quote { method }
        })[0];
        let func: TypedExpr = method.as_typed_expr();
        quote {
            pub fn bar() -> u32 {
                // Safety: test program
                unsafe {
                    func([1, 2, 3, 0])
                }
            }
        }
    }

    pub fn bar() -> u32 {
        // Safety: comment added by `nargo expand`
        unsafe {
            method([1, 2, 3, 0])
        }
    }

    pub fn test() {
        ()
    }
}

mod test_as_typed_expr_3 {
    pub comptime fn method<T, U, let N: u32>(xs_ys: ([T; N], U)) -> u32 {
        let (xs, _ys): ([T; N], U) = xs_ys;
        xs.len()
    }

    comptime fn foo(module: Module) -> Quoted {
        let method: FunctionDefinition = module.functions().filter(|f: FunctionDefinition| -> bool {
            f.name() == quote { method }
        })[0];
        let func: TypedExpr = method.as_typed_expr();
        quote {
            pub fn bar() -> u32 {
                // Safety: test program
                comptime {
                    func(([1, 2, 3, 0], "a"))
                }
            }
        }
    }

    pub fn bar() -> u32 {
        4
    }

    pub fn test() {
        ()
    }
}

mod test_as_typed_expr_4 {
    comptime fn foo(f: TypedExpr) -> Quoted {
        quote { f() }
    }

    fn bar() -> Field {
        1
    }

    fn baz() -> Field {
        let x: Field = { bar() };
        x
    }

    pub fn test() {
        ()
    }
}

mod test_as_typed_expr_5 {
    trait Trait {}

    impl Trait for i32 {}

    trait Packable<let N: u32> {
        fn pack(self);
    }

    pub struct Foo<T> {}

    impl<T> Packable<10> for Foo<T>
    where
        T: Trait,
    {
        fn pack(self)
        where
            T: Trait,
        {}
    }

    fn foo() {
        ()
    }
}
